Move kubelet ComponentConfig external types to `k8s.io/kubelet`

luxas opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This PR implements most of kubernetes/community#2354 for the kubelet.
The PR:
 - Moves `k8s.io/kubernetes/pkg/apis/kubeletconfig` as-is to `k8s.io/kubernetes/pkg/apis/config` as agreed
 - Moves the external types to the new staging repo `k8s.io/kubelet`, in the `k8s.io/kubelet/config/v1beta1` package.
 - Makes `k8s.io/kubernetes/pkg/apis/config/v1beta1` source the types from `k8s.io/kubelet/config/v1beta1`. The defaulting and conversion code is kept in this package as before.
 - All references to these packages have been updated.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
ref: kubernetes/community#2354

**Special notes for your reviewer**:

This PR depends on getting https://github.com/kubernetes/kubernetes/pull/67780 merged first.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
kubelet v1beta1 external ComponentConfig types are now available in the `k8s.io/kubelet` repo
```
/assign @sttts @mtaufen @liggitt

mtaufen commented:
We should double check for any links on the docs website that point to the things we are moving. I don't think it should be a problem (those links should always point to release branches, not master), but worth double-checking.

mtaufen commented on cmd/kubeadm/app/componentconfigs/defaults.go:
```diff
@@ -108,7 +109,7 @@ func DefaultKubeletConfiguration(internalcfg *kubeadmapi.InitConfiguration) {
 	externalkubeletcfg.HealthzPort = utilpointer.Int32Ptr(10248)
 
-	kubeletconfigv1beta1.SetDefaults_KubeletConfiguration(externalkubeletcfg)
+	kubeletconfigmetav1beta1.SetDefaults_KubeletConfiguration(externalkubeletcfg)
```
Why do this instead of `scheme.Default(externalkubeletcfg)`?

mtaufen commented on cmd/kubeadm/app/componentconfigs/BUILD:
```diff
@@ -15,9 +15,8 @@ go_library(
         "//cmd/kubeadm/app/apis/kubeadm/v1alpha3:go_default_library",
```
I think this BUILD file needs to be regenerated.
```
W0810 18:56:11.947] GoCompile: missing strict dependencies:
W0810 18:56:11.947] 	/bazel-scratch/.cache/bazel/_bazel_root/e9f728bbd90b3fba632eb31b20e1dacd/sandbox/linux-sandbox/3155/execroot/__main__/cmd/kubeadm/app/componentconfigs/defaults.go: import of "k8s.io/kubernetes/pkg/kubelet/apis/config/v1beta1"
W0810 18:56:11.947] Known dependencies are:
W0810 18:56:11.948] 	k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm
W0810 18:56:11.948] 	k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha3
W0810 18:56:11.948] 	k8s.io/kubernetes/cmd/kubeadm/app/constants
W0810 18:56:11.948] 	k8s.io/kubernetes/cmd/kubeadm/app/util
W0810 18:56:11.948] 	k8s.io/kubernetes/pkg/kubelet/apis/config
W0810 18:56:11.948] 	k8s.io/kubernetes/pkg/kubelet/apis/config/validation
W0810 18:56:11.948] 	k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig
W0810 18:56:11.948] 	k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig/v1alpha1
W0810 18:56:11.948] 	k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig/validation
W0810 18:56:11.948] 	k8s.io/apimachinery/pkg/apis/meta/v1
W0810 18:56:11.948] 	k8s.io/apimachinery/pkg/runtime
W0810 18:56:11.948] 	k8s.io/apimachinery/pkg/runtime/schema
W0810 18:56:11.949] 	k8s.io/apimachinery/pkg/runtime/serializer
W0810 18:56:11.949] 	k8s.io/apimachinery/pkg/util/runtime
W0810 18:56:11.949] 	k8s.io/apimachinery/pkg/util/validation/field
W0810 18:56:11.949] 	k8s.io/kubelet/config/v1beta1
W0810 18:56:11.949] 	k8s.io/utils/pointer
W0810 18:56:11.949] 	k8s.io/apimachinery/pkg/apis/meta/v1
W0810 18:56:11.949] 	k8s.io/apimachinery/pkg/runtime
W0810 18:56:11.949] 	k8s.io/apimachinery/pkg/runtime/schema
W0810 18:56:11.949] 	k8s.io/apimachinery/pkg/runtime/serializer
W0810 18:56:11.949] 	k8s.io/apimachinery/pkg/util/runtime
W0810 18:56:11.949] 	k8s.io/apimachinery/pkg/util/validation/field
W0810 18:56:11.949] 	k8s.io/kubelet/config/v1beta1
W0810 18:56:11.949] 	k8s.io/utils/pointer
W0810 18:56:11.950] Check that imports in Go sources match importpath attributes in deps.
W0810 18:56:12.147] INFO: Elapsed time: 30.734s, Critical Path: 8.79s
W0810 18:56:12.149] INFO: 3480 processes: 3387 remote cache hit, 1 linux-sandbox, 92 local.
W0810 18:56:12.150] FAILED: Build did NOT complete successfully
```

dims commented:
/uncc

luxas committed:
Automated package reference rename

luxas commented:
https://github.com/kubernetes/kubernetes/pull/67780 has merged, this PR is rebased and ready for review and merge in time for v1.12

thockin commented:
/lgtm
/approve

Thanks - easy to review

thockin commented:
I removed the LGTM until CI passes

sttts commented on pkg/kubelet/apis/config/doc.go:
```diff
@@ -15,5 +15,6 @@ limitations under the License.
 */
 
 // +k8s:deepcopy-gen=package
+// TODO: This file should also enable openapi-gen like the other packages
```
why? this is not served by an apiserver.

luxas replied:
We might consider this a bug, but the tag is needed in order for API violations to be tracked by the `api/api-rules/violation_exceptions.list`,  which I think is something we want to do. If we get all types to be tracked by that file automatically, while not generating any extra code, fine by me but for now let's track bad stuff there for componentconfigs as well

sttts replied:
Oh really? non-blocker for me. How you like.

luxas replied:
Gonna open an issue later about this

sttts commented on pkg/kubelet/apis/config/v1beta1/defaults.go:
```diff
@@ -21,19 +21,23 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	kruntime "k8s.io/apimachinery/pkg/runtime"
+	kubeletconfigv1beta1 "k8s.io/kubelet/config/v1beta1"
+	// TODO: Cut references to k8s.io/kubernetes, eventually there should be none from this package
```
:+1:

sttts commented on pkg/kubelet/apis/config/v1beta1/doc.go:
```diff
@@ -16,7 +16,9 @@ limitations under the License.
 // +k8s:deepcopy-gen=package
 // +k8s:conversion-gen=k8s.io/kubernetes/pkg/kubelet/apis/config
-// +k8s:openapi-gen=true
+// +k8s:conversion-gen-external-types=k8s.io/kubelet/config/v1beta1
```
is this necessary to point to itself?

luxas replied:
it's not pointing to itself, it's pointing to the staging types from the internal k8s.io meta pkg

sttts replied:
oh, not itself. nvmd.

sttts commented on staging/src/k8s.io/kubelet/CONTRIBUTING.md:
```diff
@@ -0,0 +1,7 @@
+
+Do not open pull requests directly against this repository, they will be ignored. Instead, please open pull requests against [kubernetes/kubernetes](https://git.k8s.io/kubernetes/).  Please follow the same [contributing guide](https://git.k8s.io/kubernetes/CONTRIBUTING.md) you would follow for any other pull request made to kubernetes/kubernetes.
+
+This repository is published from [kubernetes/kubernetes/staging/src/k8s.io/kube-proxy](https://git.k8s.io/kubernetes/staging/src/k8s.io/kube-proxy) by the [kubernetes publishing-bot](https://git.k8s.io/publishing-bot).
```
kubelet

sttts commented on staging/src/k8s.io/kubelet/README.md:
```diff
@@ -0,0 +1,16 @@
+# kubelet
+
+Implements [KEP 14 - Moving ComponentConfig API types to staging repos](https://github.com/luxas/community/blob/master/keps/sig-cluster-lifecycle/0014-20180707-componentconfig-api-types-to-staging.md#kubelet-changes)
```
do we really need this link?

sttts commented on staging/src/k8s.io/kubelet/config/v1beta1/doc.go:
```diff
@@ -0,0 +1,21 @@
+*/
+
+// +k8s:deepcopy-gen=package
+// +k8s:openapi-gen=true
```
not sure we need this, probably we don't. This is not served.

sttts replied:
like the other comment. If this is needed, for api violations, leave it.

luxas replied:
yeah needed for api violations for now

sttts commented:
Lgtm, some nits and the README.

luxas committed:
Move the kubelet's external types to k8s.io/kubelet

luxas committed:
Add skeleton new repo files

luxas committed:
autogenerated go code, godeps, bazel and gofmt

luxas commented:
I'm re-applying the LGTM label, as I now rebased the PR (the need for a rebase was the cause of the failing tests retried by fejta-bot).
All tests are green, and I fixed the nit in the README.
